go/arrow/ipc/message.go

```diff
@@ -0,0 +1,197 @@
+	if v, ok := flatbuf.EnumNamesMessageHeader[byte(m)]; ok {
+		return v
+	}
+	return fmt.Sprintf("MessageType(%d)", int(m))
+}
+
+const (
+	// maxNestingDepth is an arbitrary value to catch user mistakes.
+	// For deeply nested schemas, it is expected the user will indicate
+	// explicitly the maximum allowed recursion depth.
+	maxNestingDepth = 64
+)
+
+// Message is an IPC message, including metadata and body.
+type Message struct {
+	refCount int64
+	msg      *flatbuf.Message
+	meta     *memory.Buffer
+	body     *memory.Buffer
+}
+
+func NewMessage(meta, body *memory.Buffer) *Message {
+	meta.Retain()
+	body.Retain()
+	return &Message{
+		refCount: 1,
+		msg:      flatbuf.GetRootAsMessage(meta.Bytes(), 0),
+		meta:     meta,
+		body:     body,
+	}
+}
+
+// Retain increases the reference count by 1.
+// Retain may be called simultaneously from multiple goroutines.
+func (msg *Message) Retain() {
+	atomic.AddInt64(&msg.refCount, 1)
+}
+
+// Release decreases the reference count by 1.
+// Release may be called simultaneously from multiple goroutines.
+// When the reference count goes to zero, the memory is freed.
+func (msg *Message) Release() {
+	debug.Assert(atomic.LoadInt64(&msg.refCount) > 0, "too many releases")
+
+	if atomic.AddInt64(&msg.refCount, -1) == 0 {
+		msg.meta.Release()
+		msg.body.Release()
+		msg.msg = nil
+		msg.meta = nil
+		msg.body = nil
+	}
+}
+
+func (msg Message) Version() MetadataVersion {
+	return MetadataVersion(msg.msg.Version())
+}
+
+func (msg Message) Type() MessageType {
+	return MessageType(msg.msg.HeaderType())
+}
+
+func (msg Message) BodyLen() int64 {
+	return msg.msg.BodyLength()
+}
+
+// MessageReader reads messages from an io.Reader.
+type MessageReader struct {
+	r io.Reader
+
+	refCount int64
+	msg      *Message
+}
+
+// NewMessageReader returns a reader that reads messages from an input stream.
+func NewMessageReader(r io.Reader) (*MessageReader, error) {
+	return &MessageReader{r: r, refCount: 1}, nil
+}
+
+// Retain increases the reference count by 1.
+// Retain may be called simultaneously from multiple goroutines.
+func (r *MessageReader) Retain() {
+	atomic.AddInt64(&r.refCount, 1)
+}
+
+// Release decreases the reference count by 1.
+// When the reference count goes to zero, the memory is freed.
+// Release may be called simultaneously from multiple goroutines.
+func (r *MessageReader) Release() {
+	debug.Assert(atomic.LoadInt64(&r.refCount) > 0, "too many releases")
+
+	if atomic.AddInt64(&r.refCount, -1) == 0 {
+		if r.msg != nil {
+			r.msg.Release()
+			r.msg = nil
+		}
+	}
+}
+
+// Message returns the current message that has been extracted from the
+// underlying stream.
+// It is valid until the next call to Message.
+func (r *MessageReader) Message() (*Message, error) {
+	var buf = make([]byte, 4)
+	_, err := io.ReadFull(r.r, buf)
+	if err != nil {
+		return nil, errors.Wrap(err, "arrow/ipc: could not read message length")
+	}
+	msgLen := binary.LittleEndian.Uint32(buf)
```

Comment:
The spec says that the `metadata_size` is an `int32`. Isn't it a problem? (not sure)

Reply:
you're right. I'll fix that (shouldn't be a problem, though.)